go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -154,76 +285,268 @@ func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bo
+	defer npw.serviceInfoLock.Unlock()
+
+	if _, exists := npw.serviceInfo[index]; !exists {
+		// Only Set this if it dosn't exist
+		npw.serviceInfo[index] = &serviceConfig{service: service, etpHostRules: etpHostRules}
+		return false
+	}
+	return true
+
+}
+
+// updateServiceInfo sets the serviceConfig for a service and returns the existing serviceConfig, if inputs are nil
+// do not update those fields, if it does not exist return nil.
+func (npw *nodePortWatcher) updateServiceInfo(index ktypes.NamespacedName, service *kapi.Service, etpHostRules *bool) (old *serviceConfig) {
+	var exists bool
+
+	npw.serviceInfoLock.Lock()
+	defer npw.serviceInfoLock.Unlock()
+
+	if old, exists = npw.serviceInfo[index]; exists {
+		klog.V(5).Infof("No serviceConfig found for service %s in namespace %s", index.Name, index.Namespace)
+		return nil
+	}
+
+	old = npw.serviceInfo[index]
+
+	if service != nil {
+		npw.serviceInfo[index].service = service
+	}
+
+	if etpHostRules != nil {
+		npw.serviceInfo[index].etpHostRules = *etpHostRules
+	}
+
+	return old
+}
+
+// addServiceRules ensures the correct iptables rules and OpenFlow physical
+// flows are programmed for a given configuration
+func (npw *nodePortWatcher) addServiceRules(service *kapi.Service, hasHostNet bool) {
+	if hasHostNet && util.ServiceExternalTrafficPolicyLocal(service) {
+		npw.updateServiceFlowCache(service, true, hasHostNet)
+		npw.ofm.requestFlowSync()
+		addSharedGatewayIptRules(service, hasHostNet)
+	} else {
+		npw.updateServiceFlowCache(service, true, hasHostNet)
+		npw.ofm.requestFlowSync()
+		addSharedGatewayIptRules(service, hasHostNet)
+	}
+}
+
+// delServiceRules ensures the correct iptables rules and OpenFlow physical
+// flows are deleted for a given previous configuration
+func (npw *nodePortWatcher) delServiceRules(service *kapi.Service, hasHostNet bool) {
+	if hasHostNet && util.ServiceExternalTrafficPolicyLocal(service) {
+		npw.updateServiceFlowCache(service, false, hasHostNet)
+		npw.ofm.requestFlowSync()
+		delSharedGatewayIptRules(service, hasHostNet)
+	} else {
+		npw.updateServiceFlowCache(service, false, hasHostNet)
+		npw.ofm.requestFlowSync()
+		delSharedGatewayIptRules(service, hasHostNet)
+	}
+}
+
 // AddService handles configuring shared gateway bridge flows to steer External IP, Node Port, Ingress LB traffic into OVN
 func (npw *nodePortWatcher) AddService(service *kapi.Service) {
-
+	name := ktypes.NamespacedName{Namespace: service.Namespace, Name: service.Name}
 	// don't process headless service or services that doesn't have NodePorts or ExternalIPs
 	if !util.ServiceTypeHasClusterIP(service) || !util.IsClusterIPSet(service) {
 		return
 	}
-	npw.updateServiceFlowCache(service, true)
-	npw.ofm.requestFlowSync()
-	addSharedGatewayIptRules(service)
+
+	ep, err := npw.watchFactory.GetEndpoint(service.Namespace, service.Name)
+	if err != nil {
+		klog.V(5).Infof("No endpoint found for service %s in namespace %s", service.Name, service.Namespace)
+		return
```

Comment:
if there is no endpoint, that means you still need to use this function to add the right flows to get the packet into OVN for reject action. Otherwise this method is useless, because AddEndpoints will get the event. You dont need to call AddEndpoints here.